src/main/java/edu/cornell/kfs/fp/businessobject/ScheduledAccountingLineUtil.java

```diff
@@ -0,0 +1,114 @@
+package edu.cornell.kfs.fp.businessobject;
+
+import static org.kuali.kfs.sys.document.validation.impl.AccountingDocumentRuleBaseConstants.ERROR_PATH.DOCUMENT_ERROR_PREFIX;
+
+import java.sql.Date;
+import java.util.Calendar;
+import java.util.TreeMap;
+
+import org.apache.commons.lang.StringUtils;
+import org.kuali.kfs.sys.businessobject.AccountingLine;
+import org.kuali.rice.core.api.util.type.KualiDecimal;
+import org.kuali.rice.krad.util.GlobalVariables;
+
+import edu.cornell.kfs.sys.CUKFSConstants;
+import edu.cornell.kfs.sys.CUKFSKeyConstants;
+
+public class ScheduledAccountingLineUtil {
+	public static void copyFrom(ScheduledSourceAccountingLine toLine, AccountingLine other) {
+		if (ScheduledSourceAccountingLine.class.isAssignableFrom(other.getClass())) {
+			ScheduledSourceAccountingLine fromLine = (ScheduledSourceAccountingLine) other;
+			if (toLine != fromLine) {
+				toLine.setPartialAmount(fromLine.getPartialAmount());
+				toLine.setPartialTransactionCount(fromLine.getPartialTransactionCount());
+				toLine.setStartDate(fromLine.getStartDate());
+				toLine.setEndDate(fromLine.getEndDate());
+				toLine.setScheduleType(fromLine.getScheduleType());
+				toLine.setAmount(fromLine.getAmount());
+			}
+		}
+	}
+
+	public static TreeMap<Date, KualiDecimal> generateDatesAndAmounts(String scheduleType, Date startDate,
+			Date endDate, int count, KualiDecimal totalAmount, KualiDecimal partialAmount, Integer rowId) {
+		TreeMap<Date, KualiDecimal> datesAndAmounts = new TreeMap<Date, KualiDecimal>();
+
+		endDate = generateEndDate(startDate, count, scheduleType);
+
+		if (totalAmount.isNegative()) {
+			// flip the total amount to generate the entries
+			totalAmount = totalAmount.abs();
+		}
+		Date currentReversalDate = (Date) startDate.clone();
+		int i = 0;
+		while (!currentReversalDate.after(endDate) && totalAmount.isGreaterThan(KualiDecimal.ZERO)) {
+			// int i=0;
+			Calendar cRD = Calendar.getInstance();
+			cRD.setTime(startDate);
+			cRD = incrementDateByValue(scheduleType, cRD, i);
+			Date rDate = new Date(cRD.getTimeInMillis());
+			if (partialAmount.isGreaterThan(totalAmount)) {
+				datesAndAmounts.put(rDate, totalAmount);
+			} else {
+				datesAndAmounts.put(rDate, partialAmount);
+			}
+			currentReversalDate = new Date(cRD.getTimeInMillis());
+			totalAmount = totalAmount.subtract(partialAmount);
+			if (currentReversalDate.after(endDate)) {
```

Comment:
maybe not a big deal, but could you rearrange some of this logic to avoid putting the date and amount in the collection in the first place rather than removing it after the fact?